src/Features/CSharp/Portable/Diagnostics/Analyzers/CSharpTypingStyleDiagnosticAnalyzerBase.cs

```diff
@@ -0,0 +1,272 @@
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Microsoft.CodeAnalysis.CSharp.Diagnostics.TypingStyles
+{
+    internal abstract class CSharpTypingStyleDiagnosticAnalyzerBase : DiagnosticAnalyzer, IBuiltInAnalyzer
+    {
+        protected enum TypingStyles
+        {
+            None = 0,
+            VarForIntrinsic = 1 << 0,
+            NoVarForIntrinsic = 1 << 1,
+            VarWhereApparent = 1 << 2,
+            NoVarWhereApparent = 1 << 3,
+            VarWherePossible = 1 << 4,
+            NoVarWherePossible = 1 << 5
+        }
+
+        private readonly DiagnosticDescriptor _typingStyleDescriptor;
+
+        public CSharpTypingStyleDiagnosticAnalyzerBase(DiagnosticDescriptor descriptor)
+        {
+            _typingStyleDescriptor = descriptor;
+        }
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics =>
+            ImmutableArray.Create(_typingStyleDescriptor);
+
+        public DiagnosticAnalyzerCategory GetAnalyzerCategory() =>
+            DiagnosticAnalyzerCategory.SemanticSpanAnalysis;
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.RegisterSyntaxNodeAction(HandleVariableDeclaration, SyntaxKind.VariableDeclaration, SyntaxKind.ForEachStatement);
+        }
+
+        protected abstract bool IsStylePreferred(SyntaxNode declarationStatement, SemanticModel semanticModel, OptionSet optionSet, CancellationToken cancellationToken);
+        protected abstract bool AnalyzeVariableDeclaration(TypeSyntax typeName, SemanticModel semanticModel, OptionSet optionSet, CancellationToken cancellationToken, out TextSpan issueSpan);
+        protected abstract bool AnalyzeAssignment(SyntaxToken identifier, TypeSyntax typeName, EqualsValueClauseSyntax initializer, SemanticModel semanticModel, OptionSet optionSet, CancellationToken cancellationToken);
+
+        protected TypingStyles GetCurrentTypingStylePreferences(OptionSet optionSet)
+        {
+            var stylePreferences = TypingStyles.None;
+
+            stylePreferences |= optionSet.GetOption(CSharpCodeStyleOptions.UseVarForIntrinsicTypes)
+                             ? TypingStyles.VarForIntrinsic
+                             : TypingStyles.NoVarForIntrinsic;
+
+            stylePreferences |= optionSet.GetOption(CSharpCodeStyleOptions.UseVarWhenTypeIsApparent)
+                             ? TypingStyles.VarWhereApparent
+                             : TypingStyles.NoVarWhereApparent;
+
+            stylePreferences |= optionSet.GetOption(CSharpCodeStyleOptions.UseVarWherePossible)
+                             ? TypingStyles.VarWherePossible
+                             : TypingStyles.NoVarWherePossible;
+
+            return stylePreferences;
+        }
+
+        protected bool IsTypeApparentInDeclaration(SyntaxNode declarationStatement, SemanticModel semanticModel, TypingStyles stylePreferences, CancellationToken cancellationToken)
```

Comment:
Could you add a comment that defines when a type is "apparent"? I'm a bit confused by what that means -- especially because of the foreach rule below.